File: frontend/server/contexts/user/domain/value-objects/username.ts
export interface Username {
  value: string;
}

export function defineUsername(value: string): Username {
  if (!validate(value)) {
    throw new Error("Invalid username");
  }

  return {
    value,
  };
}

function validate(value: string) {
  const minLength = 6;

  return value.length >= minLength;
}

File: server/internal/user/domain/value_objects/username.go
package valueobjects

import "errors"

var (
	ErrInvalidUsername = errors.New("invalid username")
)

type Username string

func (ov Username) String() string {
	return string(ov)
}

func ValidateUsername(ov Username) error {
	return nil
	// return fmt.Errorf("invalid username field: %q", ov)
}

func NewUsername(ov Username) (*Username, error) {
	if err := ValidateUsername(ov); err != nil {
		return nil, err
	}

	return &ov, nil
}
